src/lava/lib/optimization/solvers/generic/solver.py

```diff
@@ -308,15 +328,37 @@ def _get_requirements_and_protocol(
         """
         return [CPU] if backend in CPUS else [Loihi2NeuroCore], LoihiProtocol
 
-    def _get_cost_tracking(self):
-        if self._cost_tracker is None:
+    def _get_probing(
+        self, config: SolverConfig()
+    ) -> ty.Tuple[np.ndarray, np.ndarray]:
+        """
+        Return the cost and state timeseries if probed.
+
+        Parameters
+        ----------
+        config: SolverConfig
+            Solver configuraiton used. Refers to SolverConfig documentation.
+        """
+        cost_timeseries = self._get_probed_data(
+            tracker=self._cost_tracker, var_name="optimality"
+        )
+        state_timeseries = self._get_probed_data(
+            tracker=self._state_tracker, var_name="variable_assignment"
+        )
+        if state_timeseries is not None:
+            state_timeseries &= 0x1F
+            state_timeseries = state_timeseries.astype(np.int8) >> 4

```

Comment:
Is this dublicate code? This transformation seems to happen in solution_readout/models.py as well. Is it necessary here again? Seems too low level for me for solver.py.

Reply:
Hm...
Adding such low-level detail here looks very error prone to me. Let's assume we change something about the solver (e.g. not the 5th, but the 8th position is suddenly the value at cost readout). We will probably forget to change it here.
Thus, maybe we can introduce a util function that translates between read states and the value; or an enum that defines at what position the value is. What do you think?

Reply:
Where would you place this function? What about in the models.py file of SolutionReadout? We already a member function there with the same purpose, should we make it a static method and re-use it in solver.py?